src/acsets/Columns.jl

```diff
@@ -0,0 +1,393 @@
+of `T`. Calling `add_mapping!` and `remove_mapping!` on elements out of this definable
+domain will error.
+
+However, a `InverseCache` is always defined on all of its definable domain; it defaults
+to the empty set.
+
+`InverseCache` also has a notion of "stored" domain, where inverses are actually stored.
+It is only when the inverse is stored that referential equality of inverse sets will be
+preserved, and additionally storage can have performance implications.
+
+`InverseCache`s should support the following functions:
+- [`inverse`](@ref)
+- [`add_mapping!`](@ref)
+- [`remove_mapping!`](@ref)
+- [`store!`](@ref)
+- [`maybe_unstore!`](@ref)
+- [`set_definable!`](@ref)
+"""
+abstract type InverseCache{S,T} end
+
+"""
+Arguments:
+- dom::Iterable{S}
+- m::Mapping{S,T}
+- i::InverseCache{S,T}
+- y::T
+
+Returns an iterable of the values in the domain that map onto `y`.
+"""
+function inverse end
+
+"""
+Arguments:
+- dom::Iterable{S}
+- m::Mapping{S,T}
+- i::InverseCache{S,T}
+- ys::Iterable{T}
+
+Semantically, this is the same as mapping [`inverse`](@ref) over `ys`, but the
+implementation might choose to use a view of an internal data structure of the
+index instead.
+"""
+function inverse_multi(dom, m::Mapping, i::InverseCache, ys)
+  broadcast(ys) do y
+    inverse(dom, m, i, y)
+  end
+end
+
+"""
+Arguments:
+- i::InverseCache{S,T}
+- y::T
+- x::S
+
+Add `x` to the preimage of `y`. If `y` is not in the precodomain of `i`, this MAY
+add `y` to the precodomain, or MAY error.
+"""
+function add_mapping! end
+
+"""
+Arguments:
+- i::InverseCache{S,T}
+- y::T
+- x::S
+
+Remove `x` from the preimage of `y`. Assumes that `y` is in the precodom of `i`.
+"""
+function remove_mapping! end
+
+"""
+Arguments:
+- i::InverseCache{S,T}
+- y::T
+
+Stores the inverse for `y`
+"""
+# function store! end
+
+"""
+Arguments:
+- i::InverseCache{S,T}
+- y::T
+
+Might unstore the inverse for `y` if the inverse is empty.
+"""
+function maybe_unstore! end
+
+"""
+Arguments:
+- i::InverseCache{S,T}
+- ys::Iterable{T}
+
+Sets the definable elements in `T`
+"""
+# isdefineable
+
+"""
+A column wraps a mapping and its inverse, and provides methods that do
+coordinated updates of both.
+
+Abstract Fields:
+- m::Mapping{S,T}
+- i::InverseCache{S,T}
+"""
+abstract type Column{S,T} end
+
+function (::Type{Col})() where {S,T,Col <: Column{S,T}}
+  (M,I) = Col.types
+  Col(M(), I())
+end
+
+Base.:(==)(c1::Column, c2::Column) = c1.m == c2.m
+
+Base.copy(c::T) where {T <: Column} = T(copy(c.m), copy(c.i))
+
+Base.view(c::Column, xs) = view(c.m, xs)
+
+Base.getindex(c::Column, x) = c.m[x]
+
+function Base.setindex!(c::Column, y, x)
+  if isdefined(c.m, x)
+    remove_mapping!(c.i, c.m[x], x)
+    maybe_unstore!(c.i, c.m[x])
+  end
+  store!(c.i, y)
+  add_mapping!(c.i, y, x)
+  c.m[x] = y
+end
+
+# This is to maintain compatibility with old convention for unique_indexing
+struct InconsistentPreimageFlag end
```

Comment:
Some real shade being thrown with this name ;) But really, this should have a more descriptive name, maybe something like `UnboxUniquelyIndexedPreimages`